Sources refactor and bug fixes

forgedhallpass opened the pull request.

ehsandeep reviewed (commented):
@forgedhallpass 

```
echo example.com | ./subfinder -s DNSDB -v
echo example.com | ./subfinder -s dnsdb -v
```

Both results into

```
[WRN] There is no source with the name: 'dnsdb'
```

forgedhallpass commented:
> @forgedhallpass
> 
> ```
> echo example.com | ./subfinder -s DNSDB -v
> echo example.com | ./subfinder -s dnsdb -v
> ```
> 
> Both results into
> 
> ```
> [WRN] There is no source with the name: 'dnsdb'
> ```

Good catch. That was the only source that was not lower case. The issue is now fixed and I've also added extra debug information to help with the troubleshooting.

Case 1 - Missing API key for the selected source:

![image](https://user-images.githubusercontent.com/13679401/185598570-c4761bbd-33b2-4f4f-bcc7-4a47714aba2b.png)

Case 2 - Bad API key:

![image](https://user-images.githubusercontent.com/13679401/185598431-6f944105-466d-4bbc-b23b-0ef9b366d637.png)

Mzack9999 commented on v2/pkg/runner/options.go:
```diff
@@ -203,27 +193,28 @@ func ParseOptions() *Options {
-	// unmarshal fields to migrate into temporary struct
-	if err := yaml.NewDecoder(fileSource).Decode(providers); isFatalErr(err) {
+func migrateToProviderConfig(source, dest string) error {
+	fileSource, err := os.Open(source)
```
Previously `source` was unmarshaled into a temporary providers struct, then marshaled to the destination. Here it seems it's not used anymore?

forgedhallpass replied:
![image](https://user-images.githubusercontent.com/13679401/185992814-d0c21531-ee39-4a95-b9f7-88754c88c96c.png)

forgedhallpass commented on v2/pkg/passive/sources.go:
```diff
@@ -39,185 +45,103 @@ import (
+		}
 	}
+
+	gologger.Debug().MsgFunc(func() string {
```
Why? The whole point of `MsgFunc` is to only evaluate code, if the verbose or debug flag was explicitly requested.

Mzack9999 replied:
Asynchronous access to referenced map accessed. Inline sequential eventuation will avoid potential future data races.

forgedhallpass replied:
I might be missing something. Where do you see asynchronous access?

Ice3man543 commented on v2/pkg/subscraping/utils.go:
```diff
@@ -19,12 +25,34 @@ func NewSubdomainExtractor(domain string) (*regexp.Regexp, error) {
+		gologger.Debug().Msgf("Cannot use the '%s' source because there was no API key/secret defined for it.", sourceName)
+		return result
+	}
+	return v[rand.Intn(length)]
```
we should seed the random number generator first or the items picked won't be random by `rand.Intn`

forgedhallpass replied:
I haven't touched this part directly. The previous code used `sliceutil.PickRandom` which does exactly this. I've just extended it and made it generic.

```go
func PickRandom(v []string) string {
	return v[rand.Intn(len(v))]
}
```

Ice3man543 replied:
sliceutil does call rand.Seed internally automatically iirc

forgedhallpass replied:
> `sliceutil` does call `rand.Seed` internally automatically iirc

Actually it doesn't, but I'll update my code, because it's a good point.

Ice3man543 commented on v2/pkg/subscraping/types.go:
```diff
@@ -20,54 +20,35 @@ type Source interface {
+	// NeedsKey returns true if the source requires an API key
+	NeedsKey() bool
+
+	AddApiKeys([]string)
```
couldn't we have the API keys stored somewhere instead of having a method that stores these values per provider? it just seems more elegant to have them stored somewhere as we had previously and picked randomly without this being done at time of initialization

forgedhallpass replied:
No, that's the whole point. Sources should be self-contained. About elegance, I believe my approach is far more elegant than the previous one, not to mention that it's significantly less error prone and easier to add new sources.

Ice3man543 replied:
i think there a few major problems with this approach. I like the AddApiKeys function but the way we implement the addition is an issue, mainly because the api keys are stored as global variables. which means if we want to use multiple  instances of source we cannot do it. Instead it should be stored as a variable in source itself and probably that'll be better. Also i see the function code related to API keys has been copy pasted a lot so maybe that is something we should fix as well

forgedhallpass replied:
The keys are now moved inside the `Source` struct as discussed. The code related to API keys is not copy-pasted, but uses a generic method that enables naming the key values individually for each source.

Ice3man543 commented on v2/pkg/subscraping/types.go:
```diff
@@ -20,54 +20,35 @@ type Source interface {
 	// Extractor is the regex for subdomains created for each domain
 	Extractor *regexp.Regexp
-	// Keys is the API keys for the application
-	Keys *Keys
```
this approach of storing and retreiving keys actually is simpler and easier to maintain compared to the generics one we have implemented now. Any specific reason for the current implementation? This seems to overcomplicate things

forgedhallpass replied:
I strongly disagree. See my comment above :)

Ice3man543 commented on v2/pkg/subscraping/types.go:
```diff
@@ -20,54 +20,35 @@ type Source interface {
 	RateLimiter ratelimit.Limiter
 }
 
-// Keys contains the current API Keys we have in store
```
same here

Ice3man543 commented on v2/pkg/passive/sources.go:
```diff
@@ -39,185 +46,102 @@ import (
-	"sonarsearch",
-	"virustotal",
-}
+var NameSourceMap = make(map[string]subscraping.Source, len(AllSources))
```
the make statement should be a part of the init function instead of being called globally

Ice3man543 commented on v2/pkg/subscraping/sources/github/github.go:
```diff
@@ -38,18 +39,21 @@ type response struct {
 // Source is the passive scraping agent
 type Source struct{}
 
+var apiKeys []string
```
this should be stored within source variable instead of as a global. and this is relevant for all sources

forgedhallpass replied:
done

Ice3man543 commented on v2/pkg/subscraping/utils.go:
```diff
@@ -32,6 +33,7 @@ func PickRandom[T any](v []T, sourceName string) T {
 		gologger.Debug().Msgf("Cannot use the '%s' source because there was no API key/secret defined for it.", sourceName)
 		return result
 	}
+	rand.Seed(time.Now().UnixNano())
```
this does not need to be in this function. you can just have an init() function and call it there once

Ice3man543 replied:
calling it each time is redundant and unncessary
